### PR TITLE
remove deprecated QBufferDataGenerator [qt6] and mesh 3D optimization

### PR DESCRIPTION
**Pull request by vcloarec**
following : https://github.com/qgis/QGIS/pull/42343#issuecomment-803475043

Work in progress : need a bit of optimization, bring some work in the parallel thread, and a bit of self review...

But soon ready for Qt 6!

### CODE REVIEW

**Comment by nyalldawson** [comment]
Welcome to the Qt6 porting party @vcloarec :partying_face:

**Comment by vcloarec** [comment]
Removing this deprecated generator leads to make our own geometry builder in another thread. That was also the opportunity to bring more work on this parallel thread to have less freeze time with big mesh, especially with temporal navigation.
But also need quite deep changes...

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 1]
```diff
@@ -66,17 +70,23 @@ static QByteArray createTerrainVertexData( const QgsTriangularMesh &mesh,
...
+static QByteArray createDatasetVertexData(
+  const QgsTriangularMesh &mesh,
+  const QgsMesh &nativeMesh,
+  const QgsVector3D &origin,
+  float vertScale,
+  const QgsMeshDataset3dGeometry::VertexData &data )
 {
   const int nVerts = mesh.vertices().count();
 
+  QVector<double> verticalMagnitude =
```
Comment (PeterPetrik) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:82]: const ?

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 2]
```diff
@@ -66,17 +70,23 @@ static QByteArray createTerrainVertexData( const QgsTriangularMesh &mesh,
...
+    QgsMeshLayerUtils::calculateMagnitudeOnVertices( nativeMesh, data.verticalGroupMetadata, data.verticalData, data.activeFaceFlagValues );
+
+  QVector<double> scalarMagnitude =
```
Comment (PeterPetrik) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:85]: const?

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 3]
```diff
@@ -132,9 +142,23 @@ static QByteArray createIndexData( const QgsTriangularMesh &mesh )
   return indexBytes;
 }
 
-
-static QByteArray createDatasetIndexData( const QgsTriangularMesh &mesh, const QgsMeshDataBlock &mActiveFaceFlagValues, int activeFaceCount )
+static QByteArray createDatasetIndexData( const QgsTriangularMesh &mesh, QgsMeshDataBlock &mActiveFaceFlagValues )
 {
+  int activeFaceCount = 0;
+
+  // First we nned to know about the cunt of active faces
```
Comment (PeterPetrik) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:149]: 2 typos I believe

**Comment by PeterPetrik** [comment]
@vcloarec thanks for taking lead in this. Can you please summarise the changes? I have a feeling that you solve multiple things in the one PR and I am struggling to understand what changes are required for which part of the PR, making more difficult to review. I see that 
- removing deprecated QBufferDataGenerator
- using bg thread to build 3d geometry
- changing some attributes to members instead of locals and building up active flag block for some function (what if not required/supported by the MDAL driver?)

thanks for clarificatiion.

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp (hunk 3 above)
Comment (roya0045) [reply src/3d/mesh/qgsmesh3dgeometry_p.cpp:149]: I hope those are typos.

**Comment by vcloarec** [comment]
Here, only one problem solved and one little optimization :
Problem: removing `QBufferDataGenerator` that was building the geometry in another thread, and was avoiding too long freezing of QGIS, mainly because buildings normals of the faces is expensive (for dataset rendering, needed each time the time range changes).

As Qt 3D does not replace `QBufferDataGenerator`, we need to implement our class to build geometry in another thread. That is mainly what is done in this PR. If locals variables become members, this is because, before these changes, we didn't need to have access to `QObjects` with the `QgsMesh3dGeometry` instance as parent after creation. Now with our new threaded builder, the logic is :
1 - Creation of `QgsMesh3dGeometry` : initialization (creation of buffers and attributes) and launch the geometry construction in another thread with the new geometry builder class. Not initializing buffers and attributes when creating the geometry leads to QGIS crash.

2 - When geometry builder has finished its work: `QgsMesh3dGeometry` get the data and fill bufffers and attributes with another function, so now we need the pointer to buffer and attributes as members of QgsMesh3dGeometry

For the active faces stuff, it is just refactoring and cleaning to avoid using pointers to `QgsMeshDataBlock` and to provide more thread-safe function. Indeed it is not necessary to use pointers as this object uses implicit sharing. And when active flag is not supported, the `QgsMeshDataBlock` returning by the provider is valid but empty. That's mean all faces are active. So no need to have a `nullptr` as `QQgsMeshDataBlock *activeFaceFlagValues` when active flag is not supported, just the one that is given by the provider is ok.

The optimization in this PR is sending the counting of active faces in the parallel thread (necessary to fill attributes and buffer). Before, it is done in the main thread during geometry initialization.

**Review by wonder-sk — APPROVED** [review]
Nice work!

Good so see that QBufferDataGenerator is gone in qt6 - it has always been a bit opaque when its code would get run by Qt 3D internals... The replacement using QtConcurrent looks good to me.

**Pull request merged**